java.lang.SecurityException: Prohibited package name on java 9

On java 9 there is a SecurityException occurs in tests. For instance on play-json build(in tests): 

```
play.api.libs.json.JsResultSpec *** ABORTED ***
  java.lang.ExceptionInInitializerError:
  at com.fasterxml.jackson.databind.ObjectMapper.<init>(ObjectMapper.java:581)
  at com.fasterxml.jackson.databind.ObjectMapper.<init>(ObjectMapper.java:480)
  at play.api.libs.json.jackson.JacksonJson$.<init>(JacksonJson.scala:224)
  at play.api.libs.json.jackson.JacksonJson$.<clinit>(JacksonJson.scala)
  at play.api.libs.json.StaticBinding$.generateFromJsValue(StaticBinding.scala:23)
  at play.api.libs.json.Json$.stringify(Json.scala:178)
  at play.api.libs.json.JsResult$Exception.<init>(JsResult.scala:250)
  at play.api.libs.json.JsResult$.$anonfun$toTry$default$2$1(JsResult.scala:268)
  at play.api.libs.json.JsResult$.toTry(JsResult.scala:269)
  at play.api.libs.json.JsResultSpec.$anonfun$new$5(JsResultSpec.scala:32)
  ...
  Cause: java.lang.SecurityException: Prohibited package name: java.sql
  at java.base/java.lang.ClassLoader.preDefineClass(ClassLoader.java:889)
  at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1005)
  at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
  at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:545)
  at java.base/java.net.URLClassLoader.access$100(URLClassLoader.java:83)
  at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:453)
  at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:447)
  at java.base/java.security.AccessController.doPrivileged(Native Method)
  at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:446)
  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:563)
  ...
java.lang.ExceptionInInitializerError
java.lang.ExceptionInInitializerError
	at com.fasterxml.jackson.databind.ObjectMapper.<init>(ObjectMapper.java:581)
	at com.fasterxml.jackson.databind.ObjectMapper.<init>(ObjectMapper.java:480)
	at play.api.libs.json.jackson.JacksonJson$.<init>(JacksonJson.scala:224)
	at play.api.libs.json.jackson.JacksonJson$.<clinit>(JacksonJson.scala)
	at play.api.libs.json.StaticBinding$.generateFromJsValue(StaticBinding.scala:23)
	at play.api.libs.json.Json$.stringify(Json.scala:178)
	at play.api.libs.json.JsResult$Exception.<init>(JsResult.scala:250)
	at play.api.libs.json.JsResult$.$anonfun$toTry$default$2$1(JsResult.scala:268)
	at play.api.libs.json.JsResult$.toTry(JsResult.scala:269)
	at play.api.libs.json.JsResultSpec.$anonfun$new$5(JsResultSpec.scala:32)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:1078)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at org.scalatest.WordSpec.withFixture(WordSpec.scala:1881)
	at org.scalatest.WordSpecLike.invokeWithFixture$1(WordSpecLike.scala:1076)
	at org.scalatest.WordSpecLike.$anonfun$runTest$1(WordSpecLike.scala:1088)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.WordSpecLike.runTest(WordSpecLike.scala:1088)
	at org.scalatest.WordSpecLike.runTest$(WordSpecLike.scala:1070)
	at org.scalatest.WordSpec.runTest(WordSpec.scala:1881)
	at org.scalatest.WordSpecLike.$anonfun$runTests$1(WordSpecLike.scala:1147)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:373)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:410)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.WordSpecLike.runTests(WordSpecLike.scala:1147)
	at org.scalatest.WordSpecLike.runTests$(WordSpecLike.scala:1146)
	at org.scalatest.WordSpec.runTests(WordSpec.scala:1881)
	at org.scalatest.Suite.run(Suite.scala:1147)
	at org.scalatest.Suite.run$(Suite.scala:1129)
	at org.scalatest.WordSpec.org$scalatest$WordSpecLike$$super$run(WordSpec.scala:1881)
	at org.scalatest.WordSpecLike.$anonfun$run$1(WordSpecLike.scala:1192)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.WordSpecLike.run(WordSpecLike.scala:1192)
	at org.scalatest.WordSpecLike.run$(WordSpecLike.scala:1190)
	at org.scalatest.WordSpec.run(WordSpec.scala:1881)
	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:507)
	at mill.scalaworker.ScalaWorker.$anonfun$runTests$5(ScalaWorker.scala:285)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
	at mill.scalaworker.ScalaWorker.$anonfun$runTests$2(ScalaWorker.scala:279)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at mill.scalaworker.ScalaWorker.$anonfun$runTests$1(ScalaWorker.scala:269)
	at mill.modules.Jvm$.inprocess(Jvm.scala:123)
	at mill.scalaworker.ScalaWorker.runTests(ScalaWorker.scala:264)
	at mill.scalaworker.ScalaWorker$.main(ScalaWorker.scala:70)
	at mill.scalaworker.ScalaWorker.main(ScalaWorker.scala)
Caused by: java.lang.SecurityException: Prohibited package name: java.sql
	at java.base/java.lang.ClassLoader.preDefineClass(ClassLoader.java:889)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1005)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:545)
	at java.base/java.net.URLClassLoader.access$100(URLClassLoader.java:83)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:453)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:447)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:446)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:563)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:550)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:496)
	at com.fasterxml.jackson.databind.ser.std.StdJdkSerializers.all(StdJdkSerializers.java:55)
	at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.<clinit>(BasicSerializerFactory.java:93)
	... 65 more
```

Details are here: https://travis-ci.org/rockjam/play-json/jobs/356007000#L869

Same build works fine on java 8

--- a/build.sc
+++ b/build.sc
@@ -25,6 +25,9 @@ trait MillPublishModule extends PublishModule{
       Developer("lihaoyi", "Li Haoyi","https://github.com/lihaoyi")
     )
   )
+
+  def javacOptions = Seq("-source", "1.8", "-target", "1.8")
+
 }
 object moduledefs extends MillPublishModule{
   def ivyDeps = Agg(

--- a/core/src/mill/util/ClassLoader.scala
+++ b/core/src/mill/util/ClassLoader.scala
@@ -5,22 +5,33 @@ import java.net.{URL, URLClassLoader}
 import io.github.retronym.java9rtexport.Export
 
 object ClassLoader {
-  def create(urls: Seq[URL],
-             parent: java.lang.ClassLoader)
-            (implicit ctx: Ctx.Home): URLClassLoader = {
-    val cl = new URLClassLoader(urls.toArray, parent)
-    if (!ammonite.util.Util.java9OrAbove) return cl
-    try {
-      cl.loadClass("javax.script.ScriptEngineManager")
-      cl
-    } catch {
-      case _: ClassNotFoundException =>
-        val path = ctx.home
-        val rtFile = new java.io.File(path.toIO, s"rt-${System.getProperty("java.version")}.jar")
-        if (!rtFile.exists) {
-          java.nio.file.Files.copy(Export.export().toPath, rtFile.toPath)
-        }
-        new URLClassLoader((urls ++ Some(rtFile.toURI.toURL)).toArray, parent)
+  def create(urls: Seq[URL], parent: java.lang.ClassLoader)(
+      implicit ctx: Ctx.Home): URLClassLoader = {
+    if (ammonite.util.Util.java9OrAbove) {
+      val platformParent =
+        if (parent == null)
+          classOf[ClassLoader]
+            .getMethod("getPlatformClassLoader")
+            .invoke(null)
+            .asInstanceOf[ClassLoader]
+        else parent
+      val cl = new URLClassLoader(urls.toArray, platformParent)
+      try {
+        cl.loadClass("javax.script.ScriptEngineManager")
+        cl
+      } catch {
+        case _: ClassNotFoundException =>
+          val path = ctx.home
+          val rtFile = new java.io.File(
+            path.toIO,
+            s"rt-${System.getProperty("java.version")}.jar")
+          if (!rtFile.exists) {
+            java.nio.file.Files.copy(Export.export().toPath, rtFile.toPath)
+          }
+          new URLClassLoader((urls ++ Some(rtFile.toURI.toURL)).toArray, parent)
+      }
+    } else {
+      new URLClassLoader(urls.toArray, parent)
     }
   }
 }

--- a/core/src/mill/util/ClassLoader.scala
+++ b/core/src/mill/util/ClassLoader.scala
@@ -6,24 +6,51 @@ import ammonite.ops._
 import io.github.retronym.java9rtexport.Export
 
 object ClassLoader {
+
   def create(urls: Seq[URL], parent: java.lang.ClassLoader)(
       implicit ctx: Ctx.Home): URLClassLoader = {
-    if (ammonite.util.Util.java9OrAbove) {
-      val platformParent =
-        if (parent == null)
-          classOf[ClassLoader]
-            .getMethod("getPlatformClassLoader")
-            .invoke(null)
-            .asInstanceOf[ClassLoader]
-        else parent
+    new URLClassLoader(
+      makeUrls(urls).toArray,
+      refinePlatformParent(parent)
+    )
+  }
 
+  def create(urls: Seq[URL],
+             parent: java.lang.ClassLoader,
+             customFindClass: String => Option[Class[_]])(
+      implicit ctx: Ctx.Home): URLClassLoader = {
+    new URLClassLoader(
+      makeUrls(urls).toArray,
+      refinePlatformParent(parent)
+    ) {
+      override def findClass(name: String): Class[_] = {
+        customFindClass(name).getOrElse(super.findClass(name))
+      }
+    }
+  }
+
+  private def refinePlatformParent(parent: java.lang.ClassLoader): ClassLoader = {
+    if (ammonite.util.Util.java9OrAbove) {
+      if (parent == null)
+        classOf[ClassLoader]
+          .getMethod("getPlatformClassLoader")
+          .invoke(null)
+          .asInstanceOf[ClassLoader]
+      else parent
+    } else {
+      parent
+    }
+  }
+
+  private def makeUrls(urls: Seq[URL])(implicit ctx: Ctx.Home): Seq[URL] = {
+    if (ammonite.util.Util.java9OrAbove) {
       val rtFile = ctx.home / s"rt-${System.getProperty("java.version")}.jar"
       if (!exists(rtFile)) {
         cp(Path(Export.export()), rtFile)
       }
-      new URLClassLoader((urls :+ rtFile.toNIO.toUri.toURL).toArray,platformParent)
+      urls :+ rtFile.toNIO.toUri.toURL
     } else {
-      new URLClassLoader(urls.toArray, parent)
+      urls
     }
   }
 }

--- a/core/src/mill/util/ClassLoader.scala
+++ b/core/src/mill/util/ClassLoader.scala
@@ -2,6 +2,7 @@ package mill.util
 
 import java.net.{URL, URLClassLoader}
 
+import ammonite.ops._
 import io.github.retronym.java9rtexport.Export
 
 object ClassLoader {
@@ -15,21 +16,12 @@ object ClassLoader {
             .invoke(null)
             .asInstanceOf[ClassLoader]
         else parent
-      val cl = new URLClassLoader(urls.toArray, platformParent)
-      try {
-        cl.loadClass("javax.script.ScriptEngineManager")
-        cl
-      } catch {
-        case _: ClassNotFoundException =>
-          val path = ctx.home
-          val rtFile = new java.io.File(
-            path.toIO,
-            s"rt-${System.getProperty("java.version")}.jar")
-          if (!rtFile.exists) {
-            java.nio.file.Files.copy(Export.export().toPath, rtFile.toPath)
-          }
-          new URLClassLoader((urls ++ Some(rtFile.toURI.toURL)).toArray, parent)
+
+      val rtFile = ctx.home / s"rt-${System.getProperty("java.version")}.jar"
+      if (!exists(rtFile)) {
+        cp(Path(Export.export()), rtFile)
       }
+      new URLClassLoader((urls :+ rtFile.toNIO.toUri.toURL).toArray,platformParent)
     } else {
       new URLClassLoader(urls.toArray, parent)
     }

--- a/core/src/mill/util/ClassLoader.scala
+++ b/core/src/mill/util/ClassLoader.scala
@@ -29,6 +29,14 @@ object ClassLoader {
     }
   }
 
+  /**
+    *  Return `ClassLoader.getPlatformClassLoader` for java 9 and above, if parent class loader is null,
+    *  otherwise return same parent class loader.
+    *  More details: https://docs.oracle.com/javase/9/migrate/toc.htm#JSMIG-GUID-A868D0B9-026F-4D46-B979-901834343F9E
+    *
+    *  `ClassLoader.getPlatformClassLoader` call is implemented via runtime reflection, cause otherwise
+    *  mill could be compiled only with jdk 9 or above. We don't want to introduce this restriction now.
+    */
   private def refinePlatformParent(parent: java.lang.ClassLoader): ClassLoader = {
     if (ammonite.util.Util.java9OrAbove) {
       if (parent == null)

--- a/main/src/mill/modules/Jvm.scala
+++ b/main/src/mill/modules/Jvm.scala
@@ -103,19 +103,16 @@ object Jvm {
                    classLoaderOverrideSbtTesting: Boolean,
                    body: ClassLoader => T)
                   (implicit ctx: Ctx.Home): T = {
+    val urls = classPath.map(_.toIO.toURI.toURL)
     val cl = if (classLoaderOverrideSbtTesting) {
       val outerClassLoader = getClass.getClassLoader
-      new URLClassLoader(classPath.map(_.toIO.toURI.toURL).toArray, mill.util.ClassLoader.create(Seq(), null)){
-        override def findClass(name: String) = {
-          if (name.startsWith("sbt.testing.")){
-            outerClassLoader.loadClass(name)
-          }else{
-            super.findClass(name)
-          }
-        }
-      }
+      mill.util.ClassLoader.create(urls.toVector, null, customFindClass = { name =>
+        if (name.startsWith("sbt.testing."))
+          Some(outerClassLoader.loadClass(name))
+        else None
+      })
     } else {
-      mill.util.ClassLoader.create(classPath.map(_.toIO.toURI.toURL).toVector, null)
+      mill.util.ClassLoader.create(urls.toVector, null)
     }
     val oldCl = Thread.currentThread().getContextClassLoader
     Thread.currentThread().setContextClassLoader(cl)

--- a/scalalib/test/resources/hello-world/core/src/Main.scala
+++ b/scalalib/test/resources/hello-world/core/src/Main.scala
@@ -1,5 +1,8 @@
 import scala.collection._
 import java.nio.file.{Files, Paths}
+import java.sql.Date
+import java.time.LocalDate
+import javax.swing.JButton
 
 import Main.{args, greeting}
 object Main0{
@@ -10,7 +13,9 @@ object Main0{
   }
 }
 object Main extends App {
-
+  new JButton("hello from javax")
+  val now = Date.valueOf(LocalDate.now())
+  println(s"Today is the date: ${now}")
   val person = Person.fromString("rockjam:25")
   val greeting = s"hello ${person.name}, your age is: ${person.age}"
   println(greeting)

--- a/scalalib/test/resources/hello-world/core/src/Main.scala
+++ b/scalalib/test/resources/hello-world/core/src/Main.scala
@@ -13,6 +13,8 @@ object Main0{
   }
 }
 object Main extends App {
+  // testing correct loading of java standard modules in java 9
+  // https://github.com/lihaoyi/mill/issues/251
   new JButton("hello from javax")
   val now = Date.valueOf(LocalDate.now())
   println(s"Today is the date: ${now}")

